### ISSUE TITLE
ONNXRuntime and OpenCV in Rust OBB

### ISSUE THREAD

**Issue by PonderForge** [ultralytics/ultralytics#17942]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hey everyone! 
I know this is an oddly specific question, but I need help implementing OBB in Rust for YOLOv1. I recently custom-trained my own model for human OBB, to replace my previous use of the base Detection model that Ultralytics provides. But when I tried to implement the newer model in Rust I noticed that the implementation should be the same as detection, yet the newer model was not detecting all the objects as the original.

However, I realized it was not a problem with the model as the Yolo Python implementation returns different results. Meaning that it was a problem with my implementation. I'm using OpenCV for fast resizing, and ONNX Runtime for fast inferencing.

So long story short, What's the best way to implement a OBB model in Rust using ONNX Runtime and OpenCV?

### Additional

I'm already using the ONNX Runtime for another model, so I really don't want to include another one. And I'm using OpenCV for resizing cause I need fast image processing.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @PonderForge, thank you for reaching out to us and for your interest in Ultralytics 🚀!

We recommend reviewing the [Ultralytics Docs](https://docs.ultralytics.com) where you'll find guidance on topics such as [exporting models to ONNX](https://docs.ultralytics.com/modes/export/) and using models across different frameworks.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug it. We’d need details such as the steps you've taken (Python code snippets, environment configurations, export procedure, etc.) to replicate and isolate the issue effectively.

If this is a ❓ Question regarding custom implementation, kindly share more context as appropriate. Examples of valuable context include your ONNX export parameters, details about how you are resizing images in Rust with OpenCV, and the exact inference code snippet you are utilizing. It will also help us if you reference the [YOLO Model Export Guide](https://docs.ultralytics.com/modes/export/#onnx) to ensure your ONNX model was exported correctly.

Meanwhile, we encourage exploring our community platforms for further assistance:
- Join our real-time discussions on [Discord](https://discord.com/invite/ultralytics) 🎧
- Share your insights or questions on [Discourse](https://community.ultralytics.com)
- Dive into Reddit discussions on our [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Please ensure you are using the latest `ultralytics` package and its dependencies by upgrading in your Python environment. This resolves many common issues with outdated versions:

```bash
pip install -U ultralytics
```

## Compatible Environments

Ultralytics models, including YOLO, can be run in the following verified environments, already configured with all dependencies (Python, CUDA, PyTorch, etc.):
- **Notebooks** with GPU support: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**: Use our [GCP Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon AWS**: Refer to our [AWS Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker**: Use our [Docker Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

You may also want to validate your ONNX model and Rust implementation with PyTorch—they can serve as baselines for testing against your Rust environment.

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing. CI tests validate the compatibility and performance of YOLO modes/tasks across macOS, Windows, and Ubuntu.

💡 Please note this is an automated response to ensure you receive preliminary guidance as swiftly as possible. An Ultralytics engineer will follow up soon to assist you further! 😊

**Comment by Y-T-G** [comment]
You can find an example here

https://github.com/ultralytics/ultralytics/tree/main/examples/YOLO-Series-ONNXRuntime-Rust

**Comment by PonderForge** [comment]
Thanks for responding! I've already looked into those examples and none of them use opencv. Which I'm pretty sure is my problem in pre-processing. Plus I'd rather not use another library. Do you have any other ideas?

**Comment by Y-T-G** [comment]
YOLOv8 and YOLO11 have the same pre-processing. So you can search for YOLOv8 Rust examples.

**Comment by PonderForge** [comment]
That's how I created my first implementation with the detector model by using YOLOv8 examples. My problem is since YOLOv8 is the same as YOLOv11, and the Detector model is basically the same as the OBB model (except for the radians) why are my detections different? I'll post an image example in a second

**Comment by Y-T-G** [comment]
What's the max confidence in the ONNX output?

**Comment by PonderForge** [comment]
In the OBB implementation, the minimum confidence is set at 0.5, the OBB output confidence is 0.84 for the one human and the max confidence for the model should be 1.

**Comment by PonderForge** [comment]
Ill post my code. 
Pre-processing:
```
let mut input_img: Mat = opencv::imgcodecs::imdecode( &opencv::core::Mat::from_slice(&body.to_vec()).unwrap(), IMREAD_COLOR).unwrap();
let resized_img: Mat = blob_from_image(&input_img, 1f64/255f64, Size::new(640, 640), Scalar::new(0.0,0.0,0.0,0.0), true, false, CV_32F).unwrap();
let input_tensor = ort::Tensor::from_array(([1usize,3,640,640], resized_img.data_typed::<f32>().unwrap())).unwrap();
```
Inferance: 
```
let output_tensor: SessionOutputs = detector.run(inputs!["images" => input_tensor].unwrap()).unwrap();
let outputs = output_tensor["output0"].try_extract_tensor::<f32>().unwrap().view().t().to_owned();
//Run post processing on Human Detector into Vector
let mut boxes: Vec<(f32, f32, f32, f32, f32)> = process_yolo_output(outputs, 640, 640);
```
Post Processing:
```
  fn process_yolo_output(output:Array<f32,IxDyn>,img_width: u32, img_height: u32) -> Vec<(f32,f32,f32,f32, f32)> {
      let mut boxes = Vec::new();
      let output = output.slice(s![..,..,0]);
      for row in output.axis_iter(Axis(0)) {
          let row:Vec<_> = row.iter().map(|x| *x).collect();
          let (class_id, prob) = row.iter().skip(4).enumerate()
              .map(|(index,value)| (index,*value))
              .reduce(|accum, row| if row.1>accum.1 { row } else {accum}).unwrap();
          if class_id == 0 {
              if prob < 0.5 {
                  continue
              }
              let xc = row[0]/640.0*(img_width as f32);
              let yc = row[1]/640.0*(img_height as f32);
              let w = row[2]/640.0*(img_width as f32);
              let h = row[3]/640.0*(img_height as f32);
              let x1 = xc - w/2.0;
              let x2 = xc + w/2.0;
              let y1 = yc - h/2.0;
              let y2 = yc + h/2.0;
              boxes.push((x1,y1,x2,y2,prob));
          }
      }
  
      let mut result = Vec::new();
      while boxes.len()>0 {
          result.push(boxes[0]);
          boxes = boxes.iter().filter(|box1| iou(&boxes[0],box1) < 0.7).map(|x| *x).collect()
      }
      return result;
  }
```

**Comment by glenn-jocher** [comment]
Thank you for sharing your code. Based on your implementation, it seems your pre-processing step isn't properly preparing the input for an OBB model. For YOLO models with OBB, the post-processing must account for an angle (theta) in addition to the bounding box coordinates and confidence. 

Specifically:
1. Ensure the ONNX model output contains an extra parameter for the angle (usually at index 4, before class probabilities).
2. Modify your `process_yolo_output` function to include this angle and handle OBB-specific format differences.

Use the Python implementation of Ultralytics' OBB model as a reference to align your Rust implementation more accurately. For more details, review our OBB documentation: [OBB Task Documentation](https://docs.ultralytics.com/tasks/obb). This should help standardize pre- and post-processing for your task.

**Comment by PonderForge** [comment]
Ah thank you, that would make sense. 
Btw, when I get the tensor after ONNX processed it, it returns the array:
[[535.9536],
  [589.3606],
  [199.0658],
  [116.189095],
  [4.7683716e-7],
  [0.005827891]],
Now, I pretty sure that the first 4 values are x-center, y-center, width and height, but earlier I was trying to take into account the angle, and I was unsure of what came after the first 4 values.
So was the last 2 variables confidence then angle, or angle then confidence? As in would I label [4.7683716e-7] as angle or [0.005827891] as angle?

**Comment by PonderForge** [comment]
After some testing, it seems that the very last value is radians since the max angle should be pi/2 and that's exactly what ONNX returns on some predictions. Yet @glenn-jocher, you said that the angle should be before class probabilities. Is my ONNX model off?
> 1. Ensure the ONNX model output contains an extra parameter for the angle (usually at index 4, before class probabilities).
> 2. Modify your `process_yolo_output` function to include this angle and handle OBB-specific format differences.

Here's an example of a return value: [5.9563093, 29.638168, 51.886223, 11.442295, 1.04904175e-5, 1.5710282]
None of the results go above 1.6 for the last value, but it does become negative making me think that the last value is radians, and that the second to last value is the confidence score. 

Thanks so far for your help

**Comment by glenn-jocher** [comment]
Thank you for your analysis! You are correct—the last value in the ONNX model output represents the angle (in radians), while the second-to-last value is the confidence score. This aligns with OBB model specifications. It seems your ONNX model's output is behaving as expected, with the angle correctly placed as the last value. Adjust your `process_yolo_output` function accordingly to handle these values as such. Let us know if you encounter further issues!

**Comment by PonderForge** [comment]
Thank you for the confirmation! However, this doesn't fix my original problem. Since the people in my example are vertical, the angle does not matter, and processing the radians would be easy. My issue is with the model's inaccuracy, it only manages to detect 1 human in the image when there are 3, with Rust. So what exactly does Python do to preprocess images? Also, I was testing the difference in OpenCV's image to tensor and rust image proc's image to tensor method, and there seems to be a large problem there, is there a well-known method for using OpenCV, to convert an image to a tensor that works exactly like python?

**Comment by glenn-jocher** [comment]
It appears the issue lies in the preprocessing step. Python uses libraries like PyTorch or NumPy alongside OpenCV to ensure accurate image-to-tensor conversion for YOLO. Specifically, it involves resizing, normalization (dividing pixel values by 255), and channel reordering (from HWC to CHW). Ensure you're replicating these steps accurately in Rust.

A well-known OpenCV method in Python to convert an image to a tensor includes:
1. Resize the image to the model's input size (e.g., 640x640).
2. Normalize pixel values to the range [0, 1].
3. Convert the image format from HWC to CHW.
4. Add a batch dimension.

Replicate this pipeline in Rust, and confirm it matches Python's preprocessing exactly. Testing with Python-preprocessed tensors in your Rust implementation may help identify if preprocessing is the root cause.

**Comment by PonderForge** [comment]
So I decided to just copy off of the Rust ONNXRuntime example itself for now until I can figure out how to do this better.

**Comment by Y-T-G** [comment]
The values in the output are same as detection models except there's an angle at the end: `x,y,w,h, cls_1_score, cls_2_scores, .., angle`

**Comment by PonderForge** [comment]
Yeah, eventually I copied the function for Detection in the rust example and then transcribed it to OpenCV. Now I just need to figure out how to use NMS for OBB.

**Comment by glenn-jocher** [comment]
To apply NMS for OBB in your Rust implementation, ensure your NMS function accounts for both the oriented bounding box coordinates and the additional angle parameter. You can adapt traditional NMS algorithms by using Intersection over Union (IoU) calculations specific to rotated boxes. For guidance, refer to the Python implementation and adapt it accordingly to align with your OBB task needs: [YOLO11 OBB Documentation](https://docs.ultralytics.com/tasks/obb/).
